### PR TITLE
Tx fee estimation edge cases bug

### PR DESCRIPTION
**Pull request by paweljakubas**
finish integration test

# Issue Number

#1740

# Overview

- [x] I have pin down the edge value characteristic in integration test (by bisection). This is different from Shelley testnet value reported by @johnalotoski  and @piotr-iohk , ie., `165281`
- [x] I have added integration test that reproduces issue
- [x] I have fixed the issue

# Comments
Running : 
```
stack test cardano-wallet-shelley:integration --ta '-m "TRANS_ESTIMATE_08 - Fee estimation edge case"'
```

My observations:
1. feeMin = 129 000
2. When estimating with payload amount, we have:
(a) payload < walletBalance - 129 000 => fee est = 129 000
(b) payload is (walletBalance - 129 000, walletBalance - 122 500) => fee est = 129 000 - (payload - 129 000)
(c) payload > walletBalance - 122 500 => fee est = walletBalance + (payload - 122 500 )

So when (a) fee estimation is correct, then when (b) it is decreasing linearly until we want to send `walletBalance - 122500`. Then it is increasing linearly in (c). I encoded those thresholds in integration test 

Investigation details: 
we have 3 cases (the last two give wrong fee):
1 the last proper one when 
```
CoinSel {inputs=[5_000_000], outputs=[4_871_000], change=[129_000]} this goes to rebalanceChangeOutputs twice 
-> branch : φ_original > δ_original && not (null (change s))
φ_original=129_000 δ_original=0 (change s)=[129_000]
-> branch : φ_dangling >= δ_original && φ_dangling > δ_dangling
φ_original=122_500 δ_original=129_000 φ_dangling=129_000 δ_dangling=122_500 (splitChange extraChng (change s))=[6_500]
So in senderPaysFee we have in output CoinSel {inputs=[5_000_000], outputs=[4_871_000], change=[]} as remFee=0
As a result, feeBalance is 5_000_000 - 4_871_000= 129_000 (OK)
```
2. First not proper (type 1):
```
CoinSel {inputs=[5_000_000], outputs=[4_871_001], change=[128_999]} this goes to rebalanceChangeOutputs twice 
-> branch : φ_original > δ_original && not (null (change s))
φ_original=129_000 δ_original=0 (change s)=[128_999]
-> branch : φ_dangling >= δ_original && φ_dangling > δ_dangling
φ_original=122_500 δ_original=128_999 φ_dangling=129_000 δ_dangling=122_500 (splitChange extraChng (change s))=[6_499]
So in senderPaysFee we have in output CoinSel {inputs=[5_000_000], outputs=[4_871_001], change=[]} as remFee=0
As a result, feeBalance is 5_000_000 - 4_871_000= 128_999 (NOT OK as minFee is 129_000)
```
3. then it linearly give underevaluated values (the are 6_500 cases) until we reach first case of type 2:
```
CoinSel {inputs=[5_000_000], outputs=[4_877_501], change=[122_499]} this goes to rebalanceChangeOutputs twice 
-> branch : φ_original > δ_original && not (null (change s))
φ_original=129_000 δ_original=0 (change s)=[122_499]
-> branch : φ_original > δ_original && null (change s)
φ_original=122_500 δ_original=122_499 (change s)=[]
So in senderPaysFee we have in CoinSel {inputs=[5_000_000], outputs=[4_877_501], change=[]} as remFee=1
We are trying to coverRemainingFee ! And here we do : 
-> of course no additional UTXO to be picked and we end up with exception ErrCannotCoverFee 1
As a result, we say estimating fee is 5_000_001 which is strange
```

**better picture:** 
These are value thresholds (for 1 input and 1 output) important:
```
         let (feeMin, _) = ctx ^. #_feeEstimator $ PaymentDescription                                                                                                                                          
                 { nInputs = 1                                                                                                                                                                                  
                 , nOutputs = 1                                                                                                                                                                                 
                 , nChanges = 1                                                                                                                                                                                 
                 }
``` 
```
         let (feeMin', _) = ctx ^. #_feeEstimator $ PaymentDescription                                                                                                                                          
                 { nInputs = 1                                                                                                                                                                                  
                 , nOutputs = 1                                                                                                                                                                                 
                 , nChanges = 0                                                                                                                                                                                 
                 } 
```
From so if `walletBalance - totalOuts < feeMin && walletBalance - totalOuts >= feeMin'` then linear (case 2) regime.
When, `walletBalance - totalOuts < feeMin'`   ErrCoveringFee situation (case 3)

### CODE REVIEW

**Comment by rvl** [comment]
Is it possible to replicate these results using `selectCoinsForPaymentFromUTxO` and `estimateFeeForCoinSelection` from the `Cardano.Wallet` module?

File: lib/core-integration/src/Test/Integration/Scenario/API/Shelley/Transactions.hs [hunk 1]
```diff
@@ -819,6 +819,93 @@ spec = do
...
+        wDest <- emptyWallet ctx
+        addrs <- listAddresses @n ctx wDest
+        let destination = (addrs !! 1) ^. #id
+        let (feeMin, _) = ctx ^. #_feeEstimator $ PaymentDescription
+                { nInputs = 1
+                , nOutputs = 1
+                , nChanges = 1
+                }
+
+        let amt = walletBalance - feeMin :: Natural
```
Comment (piotr-iohk) [inline lib/core-integration/src/Test/Integration/Scenario/API/Shelley/Transactions.hs:834]: :ok_hand:

**Comment by paweljakubas** [comment]
@rvl, it is even deeper, a problem is in this function https://github.com/input-output-hk/cardano-wallet/blob/master/lib/core/src/Cardano/Wallet/Primitive/Fee.hs#L217
I will prepare unit tests replicating this.

**Comment by paweljakubas** [comment]
bors try

**Comment by piotr-iohk** [comment]
I have verified on Shelley testnet and it appears to work OK.

wallet balance = 5 000 000 lovelace

estimate fee for 4834719 => 168141 lovelace
estimate fee for 4834720 => 168141 lovelace
estimate fee for 5000000 => 168141 lovelace
estimate fee for 5000001 => not_enough_money

**Comment by paweljakubas** [comment]
bors try

**Comment by paweljakubas** [comment]
bors try

**Comment by paweljakubas** [comment]
bors try

**Comment by paweljakubas** [comment]
bors try

**Comment by paweljakubas** [comment]
bors try

File: lib/core-integration/src/Test/Integration/Scenario/API/Shelley/Transactions.hs [hunk 2]
```diff
@@ -819,6 +819,100 @@ spec = do
         expectResponseCode @IO HTTP.status404 r
         expectErrorMessage (errMsg404NoWallet $ w ^. walletId) r
 
+    it "TRANS_ESTIMATE_08 - Fee estimation of edge cases" $ \ctx -> do
+        let walletBalance = 5000000
+        wSrc <- fixtureWalletWith @n ctx [walletBalance]
+        wDest <- emptyWallet ctx
+        addrs <- listAddresses @n ctx wDest
+        let destination = (addrs !! 1) ^. #id
```
Comment (KtorZ) [inline lib/core-integration/src/Test/Integration/Scenario/API/Shelley/Transactions.hs:827]: I've been fighting this "anti-pattern" from our integration DSL recently :grimacing: 
Restoring a wallet simply to get addresses that can be obtained offline is rather counter-productive. Instead, use:

- https://github.com/input-output-hk/cardano-wallet/blob/master/lib/core-integration/src/Test/Integration/Framework/DSL.hs#L1183
- https://github.com/input-output-hk/cardano-wallet/blob/master/lib/core-integration/src/Test/Integration/Framework/DSL.hs#L1208
- `mw <- entropyToMnemonic <$> genEntropy` to get a mnemonic sentence.

File: lib/core-integration/src/Test/Integration/Scenario/API/Shelley/Transactions.hs [hunk 3]
```diff
@@ -819,6 +819,100 @@ spec = do
...
+                }]
+            }|]
+        r2 <- request @ApiFee ctx
+            (Link.getTransactionFee @'Shelley wSrc) Default payload2
+        verify r2
+            [ expectSuccess
+            , expectResponseCode HTTP.status202
+            , expectField (#estimatedMin . #getQuantity)
+                (`shouldBe` feeMin)
+            ]
```
Comment (KtorZ) [inline lib/core-integration/src/Test/Integration/Scenario/API/Shelley/Transactions.hs:870]: I do not understand what this second payload and second request are testing :thinking:  ?

Comment (KtorZ) [reply lib/core-integration/src/Test/Integration/Scenario/API/Shelley/Transactions.hs:870]: Actually, I am not sure I understand what the 3rd and 4th cases below are testing either actually?

File: lib/core/src/Cardano/Wallet.hs [hunk 4]
```diff
@@ -1192,8 +1195,23 @@ estimateFeeForPayment
...
+    let totalBalance = W.balance' $ Map.toList $ W.getUTxO utxo
+    let outs = sum $ map (\(TxOut _ (Coin c)) -> c) $ NE.toList recipients
+    let wholeSel =
+            CoinSelection
+            (Map.toList $ W.getUTxO utxo)
+            (NE.toList recipients)
+            [Coin (totalBalance - outs)]
+    let feeMin = getFee $ estimateFee feeOptions wholeSel
+    if (totalBalance - outs < feeMin && totalBalance >= outs) then
+            pure $ FeeEstimation feeMin feeMin
```
Comment (KtorZ) [inline lib/core/src/Cardano/Wallet.hs:1210]: I am still not convinced than this is the _right_ fix. We are patching on top of something that should actually not happened in the first place. Plus, I don't quite understand why this assertion links together the wallet's total balance and the transaction outputs? 

What this predicates seems to suggest a particular case that isn't properly handled in the coin selection algorithm when the change output is smaller than the minimal fee.

File: lib/core-integration/src/Test/Integration/Scenario/API/Shelley/Transactions.hs [hunk 5]
```diff
@@ -819,6 +819,100 @@ spec = do
...
+            }|]
+
+        r4 <- request @ApiFee ctx
+            (Link.getTransactionFee @'Shelley wSrc) Default payload4
+        verify r4
+            [ expectSuccess
+            , expectResponseCode HTTP.status202
+            , expectField (#estimatedMin . #getQuantity)
+                (`shouldBe` feeMin)
+            ]
```
Comment (KtorZ) [inline lib/core-integration/src/Test/Integration/Scenario/API/Shelley/Transactions.hs:914]: I think this should really be captured as a unit test, integration level are slow and rather unnecessary here since this isn't an integration issue or something for which we need to observe the chain to assert about.

File: lib/core-integration/src/Test/Integration/Scenario/API/Shelley/Transactions.hs (hunk 2 above)
Comment (piotr-iohk) [reply lib/core-integration/src/Test/Integration/Scenario/API/Shelley/Transactions.hs:827]: On that one, I had a problem some time ago with this way of generating shelley addresses actually. Shelley addresses (generated with https://github.com/input-output-hk/cardano-wallet/blob/master/lib/core-integration/src/Test/Integration/Framework/DSL.hs#L1208) were actually failing in Jormungandr tests giving `400` -> see Comment on this PR https://github.com/input-output-hk/cardano-wallet/pull/1711. We were not able to find out what's the issue actually (with @paweljakubas), so went for the "anti-pattern" in the end as it was working...

File: lib/core-integration/src/Test/Integration/Scenario/API/Shelley/Transactions.hs (hunk 3 above)
Comment (paweljakubas) [reply lib/core-integration/src/Test/Integration/Scenario/API/Shelley/Transactions.hs:870]: there are two thresholds that when triggered invoke different behavior. I write in detail about it in comment of this PR. basically they are set by The two thresholds are set by : 
```
         let (feeMin, _) = ctx ^. #_feeEstimator $ PaymentDescription                                                                                                                                          
                 { nInputs = 1                                                                                                                                                                                  
                 , nOutputs = 1                                                                                                                                                                                 
                 , nChanges = 1                                                                                                                                                                                 
                 }
```
and 
```
         let (feeMin', _) = ctx ^. #_feeEstimator $ PaymentDescription                                                                                                                                          
                 { nInputs = 1                                                                                                                                                                                  
                 , nOutputs = 1                                                                                                                                                                                 
                 , nChanges = 0                                                                                                                                                                                 
                 } 
```

File: lib/core/src/Cardano/Wallet.hs (hunk 4 above)
Comment (paweljakubas) [reply lib/core/src/Cardano/Wallet.hs:1210]: right, I am also not 100% this is the best place - the place here https://github.com/input-output-hk/cardano-wallet/pull/1782 is better

File: lib/core-integration/src/Test/Integration/Scenario/API/Shelley/Transactions.hs (hunk 5 above)
Comment (paweljakubas) [reply lib/core-integration/src/Test/Integration/Scenario/API/Shelley/Transactions.hs:914]: yeah, unit testing of shelley is better as shown https://github.com/input-output-hk/cardano-wallet/pull/1782

File: lib/core-integration/src/Test/Integration/Scenario/API/Shelley/Transactions.hs (hunk 2 above)
Comment (KtorZ) [reply lib/core-integration/src/Test/Integration/Scenario/API/Shelley/Transactions.hs:827]: Shelley addresses are different from jormungandr addresses. We have 4 types of addresses: Byron, icarus, jormungandr and Shelley.

**Comment by paweljakubas** [comment]
superceded by #1782
